[Android] Correctly dispose TabbedPageRenderer

kvpt opened the pull request:
### Description of Change ###
Fix the disposing of the TabbedPageRenderer.

There are several problems with the current implementation of the dispose :
- The fragments created by the FragmentPagerAdapter was not removed
- The TabSelected listeners are not correctly cleared
- The PageChange listeners are not cleared

### Issues Resolved ### 
- Fixes #4973

### API Changes ###
 None

### Platforms Affected ### 
- Android

### Behavioral/Visual Changes ###
None

### Before/After Screenshots ### 
Not applicable

### Testing Procedure ###
- Install the repro App : [issue4973.zip](https://github.com/xamarin/Xamarin.Forms/files/2752742/issue4973.zip)
- Go to the tab of the right
- Go to another tab

### PR Checklist ###
- [X] Has automated tests <!-- (if tests are omitted or manual, state reason in description) -->
- [X] Rebased on top of the target branch at time of PR
- [X] Changes adhere to coding standard

kvpt commented:
This one can be easily covered by an UITest, I will try to make one as soon as possible.

kvpt commented:
The UI test is now included.

hartez commented:
@kvpt We've got a CI issue that's preventing us from running the automated tests - can you rebase your PR on master? That should allow us to run them.

hartez commented:
@kvpt Thanks!

kvpt commented:
@hartez, I forgot to fetch the upstream so the rebase was not complete. 
I just redo it and it must be ok now.

PureWeen commented on Xamarin.Forms.Platform.Android/AppCompat/FormsFragmentPagerAdapter.cs:
```diff
@@ -52,8 +64,28 @@ public override void RestoreState(IParcelable state, ClassLoader loader)
+
+				if (!_fragmentManager.IsDestroyed)
+				{
+					FragmentTransaction transaction = _fragmentManager.BeginTransaction();
```
```suggestion
					FragmentTransaction transaction = _fragmentManager.BeginTransactionEx();
```

kvpt commented:
I fixed it.
This bug was interesting because it led to the discovery of another problem.
The carousel page was not using the correct fragment manager when inside a navigation page, the test description help me to find that the root cause was related to that.
I also fixed the Dispose and OnElementChanged methods that did not clear all the necessary ressources.

PureWeen commented:
@kvpt can you rebase this against master? There tests are all crashing on API 19 which has been fixed on master

It looks like all the other API tests are passing with your latest change so once you rebase and we rerun the tests we should should be good

kvpt commented:
@PureWeen The rebase is done.

kvpt commented:
I rebased my branch again.
I also fixed the test 2338 which I used for testing.
Some parts of it was flaky because the event subscription was made too late.
@samhouts, @PureWeen If I have to fix remainings issues, let me know.

kvpt commented:
Fixed merge conflicts and rebased on 4.2.0.

PureWeen commented:
ui test failures unrelated all tests are passing for this PR

kvpt commented:
@samhouts Can this PR target 4.2.0 release branch like the others dispose fix. 
I have to maintain a fork specifically for these changes, as it causes crashes with no possible workaround.
Thanks.

PureWeen commented on Xamarin.Forms.Platform.Android/AppCompat/FormsFragmentPagerAdapter.cs:
```diff
@@ -52,8 +64,28 @@ public override void RestoreState(IParcelable state, ClassLoader loader)
+
+				if (!_fragmentManager.IsDestroyed)
+				{
+					FragmentTransaction transaction = _fragmentManager.BeginTransactionEx();
```
What about moving this into the override of *DestroyItem*

I tested this with the Test you have included here and it seems like it works
```C#
		public override void DestroyItem(ViewGroup container, int position, Object @object)
		{
			base.DestroyItem(container, position, @object);

			FragmentManager manager = ((Fragment)@object).FragmentManager;
			FragmentTransaction trans = manager.BeginTransactionEx();
			trans.RemoveEx((Fragment)@object);
			trans.CommitAllowingStateLossEx();
		}
```

kvpt replied:
It'a good idea, my only doubt is related to the transaction management.
It's perhaps possible to trigger a transaction overlap exception if several items are destroyed in a short period of time. 
I hope removing ExecutePendingTransactions for this case will be enough to prevent this issue.

PureWeen commented on Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Issue4973.cs:
```diff
@@ -0,0 +1,75 @@
+			RunningApp.WaitForElement(q => q.Text("Test"));
+
+			GC.Collect();
+			GC.WaitForPendingFinalizers();
```
Can you change these to use *GarbageCollectionHelper.Collect();*

kvpt replied:
Done.

PureWeen commented on Xamarin.Forms.Platform.Android/AppCompat/CarouselPageRenderer.cs:
```diff
@@ -2,16 +2,18 @@
 		bool _disposed;
 		FormsViewPager _viewPager;
 		Page _previousPage;
+		FragmentManager _fragmentManager;
```
Set this to null in dispose?

PureWeen commented on Xamarin.Forms.Platform.Android/AppCompat/FormsFragmentPagerAdapter.cs:
```diff
@@ -1,17 +1,23 @@
 		MultiPage<T> _page;
+		bool _disposed;
+
+		readonly FragmentManager _fragmentManager;
```
Set this to null in dispose?

PureWeen commented:
@kvpt the 4.3 branch UI tests are broken right now so I'd say 
- retarget this for 4.2 
- I'll kick off uitests with your latest changes
- if those pass and @samhouts wants this in 4.3 instead of 4.2 then we'll just retarget it to 4.3 and merge

kvpt commented:
@PureWeen It's done but like I say in another PR, I can't change the target branch from github.

kvpt commented:
Ok, so fast :).

PureWeen commented:
Thanks @kvpt 

I changed it and the uitests are off!

PureWeen commented on Xamarin.Forms.Platform.Android/AppCompat/CarouselPageRenderer.cs:
```diff
@@ -35,6 +37,14 @@ void ViewPager.IOnPageChangeListener.OnPageScrollStateChanged(int state)
 
 		IPageController PageController => Element as IPageController;
 
+		FragmentManager FragmentManager => _fragmentManager ?? (_fragmentManager = ((FormsAppCompatActivity)Context).SupportFragmentManager);
```
```suggestion
		FragmentManager FragmentManager => _fragmentManager ?? (_fragmentManager = Context.GetFragmentManager());
```

PureWeen replied:
I tried to just commit this myself but looks like I don't have permissions on the repository.

PureWeen commented:
```
Xamarin.Forms.Controls.Issues.BottomTabbedPageTests.AddAndRemovePages_google_pixel_2-8_1_0
```

failing

kvpt commented:
@PureWeen I tested and confirm the regression.
It's caused by the DestroyItem change, i'm not sure what cause that.
The adapter only cache three pages so when you have more than three, it can perhaps call this method on existing page that are evicted from the cache. Or perhaps we have to override another method, the 
PagerAdapter documentation is not very clear for me.
So I revert this change and rollback to dispose all the fragments at once, this way works fine.

PureWeen commented on Xamarin.Forms.Platform.Android/AppCompat/FormsFragmentPagerAdapter.cs:
```diff
@@ -31,7 +42,7 @@ public override long GetItemId(int position)
 		{
 			var fragContainer = objectValue as FragmentContainer;
-			if (fragContainer != null && fragContainer.Page != null)
+			if (fragContainer?.Page != null)
```
I think inside this code we need to remove the fragContainer from the _fragments List if it's not found. Otherwise when you remove a tab the _fragments list will still retain a references to fragContainer. 

```C#

		public override int GetItemPosition(Object objectValue)
		{
			var fragContainer = objectValue as FragmentContainer;
			if (fragContainer?.Page != null)
			{
				int index = _page.Children.IndexOf(fragContainer.Page);
				if (index >= 0)
					return index;
			}
            _fragments.Remove(fragContainer);
			return PositionNone;
		}
```

kvpt replied:
I just tested this modification.
It seem to pass tests 4973 and 1675 with success, so I think it's ok to do that.
But in this case the fragment is not removed by the FragmentManager in the dispose, don't know if it's a good thing or not. 
At least it doesn't cause crash like now.

PureWeen replied:
From my tests when adding and removing pages the fragments related to those pages get removed from the FragmentManager.

PureWeen commented:
Failing iOS test unrelated

PureWeen commented:
@kvpt for now I'm merging this to 4.3.0 because of where we are in our release cycles. I realize this one has taken awhile to get merged (apologies) but we'd rather have this merged into our prereleases for now so it can get a little more battle tested.
